### DISCUSSION TITLE
Confused with the Docker build

### DISCUSSION THREAD

**Discussion by DimitriDR** [retypeapp/retype#474] [Q&A]
Hi,

Okay, so, I have several questions.
First of all, I did build the Dockerfile like explained [here](https://retype.com/hosting/docker/). It starts perfectly, good. It just displays a single welcome page, It seems correct here.

Now, I try to map a volume from an accessible folder (containing few markdown files) to `/usr/local/apache2/htdocs/`, such: `sudo docker container run --rm -p -v /volume1/master/markdown_test/:/usr/local/apache2/htdocs/ 49194:80 name/retype`

But here, I have an error, access denied.

I see there's some kind of `retype.yml` file to give some directives to retype. But where should this file be?

And by the way, commands `retype`, `dotnet`, `yarn`, etc. are "not found".

I'm highly confused about this. Any help please?

**Comment by DannyBen** [comment]
Not sure where your problem is exactly, but I can offer some advice:

1. Do not use `sudo` with docker. If your user cannot run docker without `sudo`, add yourself to the docker group by running `sudo usermod -aG docker "$(whoami)"`
2. Before you start to apply your own additions to the instructions on the docker page, I propose you first check that it works as is. Copy the provided Dockerfile exactly, and place it **in a directory that already contains your Retype source**.
3. Build the image: `docker build -t mysite .`
4. Run it without mounting anything (since the purpose of this image is to already contain the generated documentation): `docker run --rm -it -p 3000:80 mysite`

**Comment by DannyBen** [comment]
A couple additional notes:

The instructions on the docker page, as I mentioned, are designed to provide you with a "build and serve" solution.

There are a couple more use cases that the help page does not cover:

### Just Serve

If you have already built your documentation, or plan on building it locally without docker, you can use a Dockerfile like this:

```
FROM httpd
COPY ./docs/ /usr/local/apache2/htdocs/
```

Then, you can run it with or without volume mount:

```
$ docker run --rm -p 3000:80 imagename
$ docker run --rm -p 3000:80 -v $PWD/docs:/usr/local/apache2/htdocs/ imagename
```

(assuming your rendered documentation is in `./docs` locally).

### Retype CLI in Docker

If you just wish to have the `retype` tool itself Dockerized, you can use a Dockerfile like this:

```
FROM mcr.microsoft.com/dotnet/sdk:6.0 AS builder
WORKDIR /build
RUN dotnet tool install retypeapp --tool-path /bin
ENTRYPOINT ["retype"]
```
(which is the first half of the Dockerfile in the help page + an `ENTRYPOINT`.)

then you can run it with:

```
$ docker run -it --rm imagename --version
```

**Comment by DimitriDR** [comment]
Thanks for your exhaustive reply @DannyBen 

For the `sudo`, It's on my Synology NAS, I don't want to mess with it, and I'm not sure if It represents a special danger to not need to type the root password with Docker, I'm not sure if It's a good practice to do that?

So, It perfectly works when files are in the same folder.

But what if I have a synchronized folder with Git? The best would be to just watch for new files and update the website accordingly.

I tried your last solution, I did:

```
sudo docker run --rm -it -p 3000:80 mysite init .
sudo docker run --rm -it -p 3000:80 mysite build .
sudo docker run --rm -it -p 3000:80 mysite run .
```

But suddently, It wants to run on port 5000. So I did:

`sudo docker run --rm -it -p 3000:5000 mysite run .`

But It is still not accessible. :'(

Yeah, you guessed it, I'm not good with Dockerfile.

**Comment by DimitriDR** [comment]
Okay, I tried somethings, but Docker compose doesn't set it up:

```
version: "3.9"

services:
  retype:
    image: dimitridr/retype:latest
    container_name: retype
    restart: unless-stopped
    entrypoint: retype
    command: retype watch
    ports:
      - "50100:80"

    networks:
      - retype-network

networks:
  retype-network:
    name: retype-network
    driver: bridge
```

where the image has been built with:
```

FROM mcr.microsoft.com/dotnet/sdk:6.0 AS builder
WORKDIR /build
COPY . /build
RUN dotnet tool install retypeapp --tool-path /bin
RUN retype build --output .docker-build/
ENTRYPOINT ["retype"]

FROM httpd:latest
COPY --from=builder /build/.docker-build/ /usr/local/apache2/htdocs/
```

I guess you know where I want to go. I just want to build it once, let retype watch new files, and forget it.

But problems:

1. DC doesn't set it up, with entrypoint and command removed, It doesn't work.
2. it listens on port (localhost:)5000, but even If I map it to port: whatever:5000 ,It does not work.
